aggregate_root/spec/instrumented_repostory_spec.rb

```diff
@@ -0,0 +1,86 @@
+require 'spec_helper'
+require 'active_support/notifications'
+
+module AggregateRoot
+  RSpec.describe InstrumentedRepository do
+
+    describe "#load" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository, ActiveSupport::Notifications)
+        aggregate = Order.new
+
+        instrumented_repository.load(aggregate, 'SomeStream')
+
+        expect(some_repository).to have_received(:load).with(aggregate, 'SomeStream')
+      end
+
+      specify "instruments" do
+        instrumented_repository = InstrumentedRepository.new(spy, ActiveSupport::Notifications)
+        subscribe_to("load.repository.aggregate_root") do |notification_calls|
+          aggregate = Order.new
+
+          instrumented_repository.load(aggregate, 'SomeStream')
+
+          expect(notification_calls).to eq([{
+            aggregate_class: Order,
+            stream_name: 'SomeStream',
+          }])
+        end
+      end
+    end
+
+    describe "#store" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository, ActiveSupport::Notifications)
+        aggregate = Order.new
+
+        instrumented_repository.store(aggregate, 'SomeStream')
+
+        expect(some_repository).to have_received(:store).with(aggregate, 'SomeStream')
+      end
+
+      specify "instruments" do
+        instrumented_repository = InstrumentedRepository.new(spy, ActiveSupport::Notifications)
+        subscribe_to("store.repository.aggregate_root") do |notification_calls|
+          aggregate = Order.new
+          aggregate.create
+          aggregate.expire
+
+          instrumented_repository.store(aggregate, 'SomeStream')
+
+          expect(notification_calls).to eq([{
+            aggregate_class: Order,
+            aggregate_version: -1,
```

Comment:
Interesting question, are we more interested in version before or after storing events? Also what happens in context of instrumentation if we cannot store events?

Reply:
Usually repository gets a new instance of aggregate, so on load method call version does not matter. The version in store instrumentation is an aggregate version **before** store.